### ISSUE TITLE
Connection to proxy lagging/timing out on server restarts

### ISSUE THREAD

**Issue by chrismwiggs** [GeyserMC/Floodgate#174]
While using Floodgate 2.0 on my waterfall (1.17, build 440) and Geyser standalone (b742), anytime a server under the proxy restarts, the entire network's ping jumps to 16k+ and usually disconnects 90% of players. This is something new, only happening since the Floodgate 2.0 update (we have been using this setup for a while, pre-update, with no issues).

When this happens, there are no errors on the proxy, the individual servers, or the Geyser instance. It is just like the proxy lags out; you can run commands on the servers via console, but anything that requires a connection in-game is delayed by 30s+, or you just get timed out. 

The only useful errors I have seen are that some users will get a "too many packets" error when timing out. I've seen something that suggest that is an error that can happen with ViaVersion, which we do also use on the proxy.

**Comment by chrismwiggs** [comment]
Following up back on this issue as it's still happening and likely not a Geyser issue as mention in this issue: https://github.com/GeyserMC/Geyser/issues/2399

**Comment by chrismwiggs** [comment]
Can confirm this is happening with and without global linking (thought maybe it's some sort of connection issues to an external). 

It also seems to not be happening right away after the Waterfall proxy + Geyser have been restarted. The first few paper restarts seem to be fine, but then this issue will crop up and debilitate the entire network, dropping dozens of players.

**Comment by chrismwiggs** [comment]
Following up again. Still an issue. 

Have been trying anything I can do to resolve it, but nothing is working. Tried moving Viaversion to the individual servers, still happening. Tried with and without external linking on, still happens.

Is there anyone who checks this repo? I don't think I am crazy here, this seems like something is going wrong in Floodgate

**Comment by DomDev14** [comment]
I have the exact same issue, but only geyser is installed on proxy side. I think it has todo with geyser. I think this behaviour can be mitigated to some degree when geyser as standalone jar and not using floodgate as for now

**Comment by chrismwiggs** [comment]
I've had the issue both on the proxy and on the standalone version, but I've given up on getting support on this. Haven't had even acknowledgment of this issue save for when it was on the Geyser repo (which is the only piece anyone cares about I guess). Ignored on Discord after multiple attempts too.

My only last thought is a weird java version issue; I haven't tried this all on Java 16 and am hoping it resolves it but if not, I won't be using Geyser and Floodgate as it's not worth the damage it does to an active network.

**Comment by Camotoy** [comment]
Our apologies on the lack of communication. It's difficult to diagnose these issues when developers don't get the same issue :/. Someone posted plugins that allegedly cause the issue and that just requires one of us to install MySQL. Did you try without other plugins?

DomDev, are Java players also affected?

**Comment by chrismwiggs** [comment]
I will try a setup with just Geyser and Floodgate later tonight to see and will use Java 16 because I am curious. It's very hard to troubleshoot though because it requires actual player load, with both java and bedrock players. Every time I've thought it was fixed (from testing with one or two extra players without issue), as soon as real load hits and a server restarts, it pops back up.

I think the issue is just that nothing has really been done to try and get you more information to either reproduce it or track it down. I'm happy to do whatever is needed to track it down, but it feels like because no one on the dev team has seen it, it just gets dismissed as not possible. Whatever information you need, I will do my best to get you, I just need to know what is needed!

I should also add, as I've described my issue, there is no affecting just one type of user; the entire proxy locks up. Anything going in or out is frozen as if suddenly the entire machine has ping issues with all players (the ones who manage to not get kicked/timeout!) showing ping in the 15k+ range (when usually we're seeing 20-60ms).

**Comment by Camotoy** [comment]
What about CPU usage?

**Comment by DomDev14** [comment]
Hello Camotoy, I had Geyser + floodgate installed on the proxy as plugin, but nothing on our subservers. 
proxysoftware is waterfall, serversoftware is paper. I think it definitively affeccted all players.

**Comment by Tim203** [comment]
> Following up again. Still an issue.
> 
> Have been trying anything I can do to resolve it, but nothing is working. Tried moving Viaversion to the individual servers, still happening. Tried with and without external linking on, still happens.
> 
> Is there anyone who checks this repo? I don't think I am crazy here, this seems like something is going wrong in Floodgate

If this issue isn't Bungeecord specific then I expect that https://github.com/GeyserMC/Floodgate/commit/beda917fc97086379ab9565e4e50a34034f1a918 fixes your issue.
If it is Bungee specific then I expect that we're talking about #154

**Comment by chrismwiggs** [comment]
can confirm it is not #154, we aren't talking about a minor ping increase all the time, this is a server breaking ping that is consistently 16,000+ when a server restarts (after 2 minutes it returns to normal 20-60ms). 

Just tried the latest builds of floodgate and geyser (which I assume include beda917) and the issue still happens.

**Comment by Camotoy** [comment]
If we have yet to ask, can we get a Spark report during this process? 

> Spark is a plugin that helps you monitor performance for you server.
> https://spark.lucko.me/download

> To record performance on your server use:
> `/spark profiler --thread *` . Once the monitored action is done, use `/spark profiler --stop`.

**Comment by chrismwiggs** [comment]
https://spark.lucko.me/DYFr0e5VIy Here is a profile from when the lockup starts, ended when the server was responding again. 

Again, the server will respond from the console; this is *not* the server being overloaded or under-resourced. This issue is also only present when using Floodgate and Geyser (this exact setup has been running fine for 18 months pre geyser). 

Interestingly, while trying to get the lockup to happen, I've found the lockup only happens when the spigot server that restarts has players joining it right away. If no one joins the server for 5-10 minutes, the other servers feel unaffected. Without knowing how any of geyser/floodgate works, that feels to me that it *is* part of the auth process (ie floodgate, maybe related to beda917, who knows.)

**Comment by Camotoy** [comment]
Something called `ProjectKorra` is creating a lot of threads. I think having a reference Spark with things working normally would be helpful.

**Comment by Tim203** [comment]
I don't see any mention of Floodgate nor Geyser in that dump

**Comment by chrismwiggs** [comment]
I don't think it is particularly helpful, if you are looking for things to blame for a performance hit. Like I have said, this issue does not occur without floodgate or geyser and the load from Projectkorra is normal. The database queries are also normal, and are not causing the entire network to lock up _only_ with geyser/floodgate (neither of which are using the database). 

_Again_, this is not the spigot servers locking up, running out of resources or otherwise freezing (you can interact with them directly). If there is some tool to profile bungee while this happens, I suspect we will see something more useful there. Do you have any reccomendations? If you really need to see a Spark without geyser and floodgate, I can provide one, but it is going to be largely the same for all the reason I've provided.

**Comment by Tim203** [comment]
The problem is that there is no mention of Floodgate nor Geyser in that profile, which makes me question if they are even installed.

**Comment by Camotoy** [comment]
I think he's trying to say that the backend servers freeze up? While Floodgate is on a proxy and Geyser is standalone. It doesn't seem logical but here we were.

**Comment by Tim203** [comment]
If that's the case then it's not a Floodgate issue, it's just not possible. Random CPU spikes (which can lead to increased ping) on Bungee and forks is possible and confirmed, but then this issue becomes a duplicate of #154

**Comment by chrismwiggs** [comment]
> The problem is that there is no mention of Floodgate nor Geyser in that profile, which makes me question if they are even installed.

Of course they are not installed; this profile is from the spigot server; floodgate is on bungee and geyser is running as a standalone instance. There is no possible way for them to appear on the profile of the spigot server. 



> I think they're trying to say that the backend servers freeze up? While Floodgate is on a proxy and Geyser is standalone. It doesn't seem logical but here we are.

Define backend? I think there is something being lost in translation here. Spigot servers do not freeze, bungee server does freeze. The standalone geyser server does not freeze. 



> If that's the case then it's not a Floodgate issue, it's just not possible. Random CPU spikes (which can lead to increased ping temporarily) on Bungee and forks is possible and confirmed, but then this issue becomes a duplicate of #154

I don't see how this is the same issue. After this initial lockup, ping returns to 20-60ms for all players (bedrock and java). It seems like #154 is describing overall higher ping (and I assume higher cpu usage) consistently. Again, after the initial lockup, everything works as expected (and as it does 100% of the time without floodgate and geyser). If this is being blamed on CPU spikes, this is a much bigger problem; this is a 3950x running 7 servers in dockerized containers; if floodgate/geyser can chew through that, something is very wrong. To be clear, I am not reporting lag or increased resource usage; this is specifically an issue where a server reconnects to the proxy (a proxy running with 8gb ram and 2 dedicated cores on a 3950x, which should hardly be a cpu issue)

Floodgate IS on the proxy and it IS the proxy that locks up, so I don't know how this isn't possible. I feel like at this point you think I am just making it up; if you want a demonstration I am happy to do a restart while you are on the network to see exactly what the issue is in real time.

**Comment by Camotoy** [comment]
Oh, then having a Spark of the backend server isn't going to help - can you send one from the proxy?

**Comment by chrismwiggs** [comment]
Here is a profile from the proxy: https://spark.lucko.me/Se6ds6ksRJ

Another note, when this lock up is happening you can *also* run commands on the proxy (like /glist, returns instantly). Feels more like something is wrong with how players are handled vs resources (otherwise you'd not be able to do anything as a player or on the console, in theory?).

**Comment by chrismwiggs** [comment]
Great to hear! Not sure if it helps, but I've been restarting servers and putting a whitelist on them for 5 minutes after restart, which has entirely stopped the lock ups. Seems to really be about players joining the servers during what feels like an extended start up/handshake process or something.

Let me know if I can provide any other info!

**Comment by Tim203** [comment]
Hi, Floodgate build 29 has just been released which should fix this issue. Let me know if you're still having problems

**Comment by chrismwiggs** [comment]
I've updated and am cautiously reporting that I'm not seeing the lockups. Only have restarted two servers, both with only a handful of players online. Will know for sure over the next day or two and will report back. Thanks!

**Comment by Tim203** [comment]
Cool. I'll label it as resolved until we hear otherwise
